### PR TITLE
#286 Puzzle 213 in HttpFacade.java:66-69: Create new a class Facade

### PR DESCRIPTION
**Pull request by simonjenga**
#286 Puzzle 213 in HttpFacade.java:66-69: Create new a class Facade

static subclasses that encapsulate BaseFacade, please review

### CODE REVIEW

**Comment by davvd** [comment]
@simonjenga Thanks, I will ask someone to do a code review here, thanks for your pull request

**Comment by simonjenga** [comment]
@davvd @yegor256 Please assign someone to do a CR on this PR and provide some feedback.

**Comment by davvd** [comment]
@krzyk please review, thanks

File: .gitignore [hunk 1]
```diff
@@ -6,6 +6,7 @@ src/main/latex/*.pdf
 src/main/latex/*.aux
 src/main/latex/*.gz
 local-repo/
+/bin
```
Comment (krzyk) [inline .gitignore]: @simonjenga why did you add this?

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+ * Convenient base superclass that creates a request dispatcher object for port
+ * listener implementations such as {@link HttpFacade} or {@link FtpFacade}.
+ *
+ * @author Simon Njenga (simtuje@gmail.com)
+ * @author Yegor Bugayenko (yegor@tpc2.com)
+ * @version $Id$
+ * @since 0.1
+ */
+@ToString
+@EqualsAndHashCode(of = { "sockets", "server", "secured" })
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga `frontend` and `backend` is missing

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        final int backthreads, final String backprefix, final ServerSocket svr,
+        final ServerSocket secure) {
+        this.frontend = this.createThreadPool(frontthreads, frontprefix);
+        this.backend = this.createThreadPool(backthreads, backprefix);
+        this.server = svr;
+        this.secured = secure;
+    }
+
+    /**
+     * {@inheritDoc}
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga we don't need any javadoc on overriden methods

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+     * Dispatcher for either HttpThread or FTPThread.
+     * @param thread The dispatcher for processing a thread
+     * @param bckend The executor service
+     */
+    private void threadRunnableDispatcher(final RequestDispatcher thread,
+        final ScheduledExecutorService bckend) {
+        final Runnable runnable = new VerboseRunnable(
+            new BaseFacade.ThreadRunnableDispatcher(thread), true, true
+        );
+        for (int idx = 0; idx < THREADS; ++idx) {
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga please add `BaseFacade.` before `THREADS` (all static fields have to be qualified with class name)

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 5]
```diff
@@ -54,26 +54,24 @@
...
  */
 @ToString
 @EqualsAndHashCode(of = { "hosts", "sockets" })
-final class FtpThread {
+final class FtpThread implements RequestDispatcher {
 
     /**
      * Queue of sockets to get from.
      */
-    @NotNull
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: @simonjenga why did you remove `NotNull`?

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 6]
```diff
@@ -54,26 +54,24 @@
...
     private final transient BlockingQueue<Socket> sockets;
 
     /**
      * Hosts to work with.
      */
-    @NotNull
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: @simonjenga why did you remove `NotNull`?

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 7]
```diff
@@ -84,10 +82,9 @@
...
 
     /**
-     * Dispatch one request from the encapsulated queue.
-     * @return Amount of bytes sent to socket
-     * @throws InterruptedException If interrupted while waiting for the queue
+     * {@inheritDoc}
      */
+    @Override
     public long dispatch() throws InterruptedException {
         assert this.hosts != null;
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: @simonjenga those asserts can be removed instead

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java [hunk 8]
```diff
@@ -84,10 +82,9 @@
     }
 
     /**
-     * Dispatch one request from the encapsulated queue.
-     * @return Amount of bytes sent to socket
-     * @throws InterruptedException If interrupted while waiting for the queue
+     * {@inheritDoc}
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: @simonjenga please remove this javadoc

File: s3auth-relay/src/main/java/com/s3auth/relay/HttpThread.java [hunk 9]
```diff
@@ -136,13 +136,12 @@
     }
 
     /**
-     * Dispatch one request from the encapsulated queue.
-     * @return Amount of bytes sent to socket
-     * @throws InterruptedException If interrupted while waiting for the queue
+     * {@inheritDoc}
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/HttpThread.java]: @simonjenga pls remove this javadoc

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+                2, "FTP-front", THREADS, "FTP-back",
+                    new ServerSocket(port), null
+            );
+            this.facade = this;
+            final FtpThread thread = new FtpThread(this.facade.sockets, hosts);
+            this.facade.threadRunnableDispatcher(thread, this.facade.backend);
+        }
+
+        /**
+         * {@inheritDoc}
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga please remove this javadoc

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+         */
+        @Override
+        public void listen() {
+            this.facade.listen(this.facade.frontend, this.facade.server);
+        }
+
+        /**
+         * {@inheritDoc}
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga please remove this javadoc

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+     * @version $Id$
+     * @since 0.1
+     * @see Main
+     * @todo #213:30min Implement TLS secure port listening in a manner analogous
+     * to {@link HttpFacade}.
+     */
+    @ToString
+    @SuppressWarnings("PMD.DoNotUseThreads")
+    @Loggable(Loggable.DEBUG)
+    protected static final class FtpFacade extends BaseFacade {
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga plase make it a top level class

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+         * {@inheritDoc}
+         */
+        @Override
+        public void listen() {
+            this.facade.listen(this.facade.frontend, this.facade.server);
+            this.facade.listen(this.facade.frontend, this.facade.secured);
+        }
+
+        /**
+         * {@inheritDoc}
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga please remove this javadoc

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+            );
+            this.facade = this;
+            final HttpThread thread = new HttpThread(
+                this.facade.sockets, hosts
+            );
+            this.facade.threadRunnableDispatcher(thread, this.facade.backend);
+        }
+
+        /**
+         * {@inheritDoc}
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga please remove this javadoc

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 15]
```diff
@@ -0,0 +1,435 @@
...
+        * @param hosts Hosts
+        * @param port Port number
+        * @param sslport SSL port number.
+        * @throws IOException If can't initialize
+        */
+        HttpFacade(@NotNull final Hosts hosts, final int port,
+            final int sslport) throws IOException {
+            super(
+                2, "front", THREADS, "back",
+                        new ServerSocket(port),
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga this line (and two lines below) have to much indentation, it can be on the same level as `2`

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 16]
```diff
@@ -0,0 +1,435 @@
...
+     * @author Simon Njenga (simtuje@gmail.com)
+     * @author Yegor Bugayenko (yegor@tpc2.com)
+     * @version $Id$
+     * @since 0.1
+     * @see Main
+     */
+    @ToString
+    @SuppressWarnings("PMD.DoNotUseThreads")
+    @Loggable(Loggable.DEBUG)
+    protected static final class HttpFacade extends BaseFacade {
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga please make it a top level class

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 17]
```diff
@@ -0,0 +1,435 @@
...
+       /**
+        * Package-private constructor.
+        * @param thrd The RequestDispatcher
+        */
+        ThreadRunnableDispatcher(final RequestDispatcher thrd) {
+            this.thread = thrd;
+        }
+
+        /**
+         * {@inheritDoc}
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga please remove this javadoc

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 18]
```diff
@@ -0,0 +1,435 @@
...
+     */
+    @ToString
+    @SuppressWarnings("PMD.DoNotUseThreads")
+    @Loggable(Loggable.DEBUG)
+    protected static final class FtpFacade extends BaseFacade {
+
+       /**
+        * Base facade encapsulated.
+        */
+        private final transient BaseFacade facade;
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga why do you inherit AND encapsulate the base class?

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 19]
```diff
@@ -0,0 +1,435 @@
...
+    /**
+     * Report overflow problem to the socket and close it.
+     *
+     * The default implementation is <em>empty</em>. This should be overridden
+     * in the subclasses as necessary.
+     *
+     * @param socket The socket to report to
+     * @throws IOException If something wrong happens inside
+     */
+    protected void overflow(final Socket socket) throws IOException {
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga this method could be abstract if you would just inherit `BaseFacade`

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 20]
```diff
@@ -0,0 +1,435 @@
...
+    }
+
+    /**
+     * Start listening to the ports.
+     *
+     * The default implementation is <em>empty</em>. This should be overridden
+     * in the subclasses as necessary.
+     */
+    protected void listen() {
+        // nothing to do, intentionally empty
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga this method couold be abstract if you would just inherit from `BaseFacade`

File: s3auth-relay/src/main/java/com/s3auth/relay/RequestDispatcher.java [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * Dispatcher for a single {@link HttpThread} or {@link FtpThread}.
+ *
+ * <p>All implementations of this interface must be thread-safe.
+ *
+ * @author Simon Njenga (simtuje@gmail.com)
+ * @version $Id$
+ * @since 0.1
+ */
+interface RequestDispatcher {
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/RequestDispatcher.java]: @simonjenga maybe this could be named just `Dispatchable`

File: s3auth-relay/src/main/java/com/s3auth/relay/RequestDispatcher.java [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.s3auth.relay;
+
+/**
+ * Dispatcher for a single {@link HttpThread} or {@link FtpThread}.
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/RequestDispatcher.java]: @simonjenga there is no need to mention `HttpThread` or `FtpThread`, interface is pretty generic

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 23]
```diff
@@ -0,0 +1,435 @@
...
+     */
+    @ToString
+    @SuppressWarnings("PMD.DoNotUseThreads")
+    @Loggable(Loggable.DEBUG)
+    protected static final class HttpFacade extends BaseFacade {
+
+       /**
+        * Base facade encapsulated.
+        */
+        private final transient BaseFacade facade;
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga same here, why it is also encapsulated?

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 24]
```diff
@@ -0,0 +1,435 @@
...
+ * listener implementations such as {@link HttpFacade} or {@link FtpFacade}.
+ *
+ * @author Simon Njenga (simtuje@gmail.com)
+ * @author Yegor Bugayenko (yegor@tpc2.com)
+ * @version $Id$
+ * @since 0.1
+ */
+@ToString
+@EqualsAndHashCode(of = { "sockets", "server", "secured" })
+@SuppressWarnings({ "PMD.DoNotUseThreads", "PMD.TooManyMethods" })
```
Comment (krzyk) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga when you move FTP and HTTP classes out, you can remove `TooManyMethods`

**Comment by krzyk** [comment]
@simonjenga please see my comments

**Comment by simonjenga** [comment]
@krzyk /bin is included in order to allow git to exclude compiled class files (*.class) from being commited. Eclipse compiles projects and puts those .class files in that bin folder.

I removed NotNull because validation has already been done at the constructor for those two fields. I don't think we need to duplicately validate them again. The constructor will never allow initialization of null values for those fields.

BaseFacade cannot be an abstract class. It will not allow encapsulation to be done. There is an error message related to Qulice when you try to do this.

#overflow(final Socket socket) and #listen() cannot be abstract methods since it would mean BaseFacade to be an abstract class which is not the case as explained above.

Base facade is encapsulated in order to access the private transient fields within it. Rememmber, we have been requested not to use getters/setters by the Architect. That is why i put those subclasses inside the Base class.

what is a top level class?

**Comment by krzyk** [comment]
@simonjenga please respond in context (where the comments were)
As for the design, you could e.g. make `BaseFacade` (but you will have to find a better name) encapsulate object that will have specific behaviors (one for FTP and one for HTTP).

e.g. (suggestion, not actual design, you will have to try it out):
```
public class BaseFacade {
    private final Protocol protocol:

    public BaseFacade(Protocol prot) {
        this.protocol = prot;
    }

    private void overflow(Socket socket) {
        this.protocol.overflow(socket);
    }

....
}

class FTP implements Protocol {.....}
class HTTP implements Protocol {....}
```

**Comment by simonjenga** [comment]
@krzyk While i try out to see if your proposed design will work, please look at this class ```com.jcabi.http.request.BaseRequest``` I have designed my solution with reference to it.

File: s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java (hunk 7 above)
Comment (simonjenga) [inline s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java]: @krzyk These asserts will be romoved when the method will be implemented. They help to prevent warnings related to unused private fields in the class.

**Comment by simonjenga** [comment]
@krzyk Not all ```{@inheritDoc}``` tags can be removed. Some are needed by qulice to prevent build errors.

Your proposed design to introduce an interface ```Protocol``` won't provide much help. We will still come across the same problems i experienced before in trying to access ```private transient``` fields and that is why our subclasses cannot be top level classes.

See the addressed code review changes.

**Comment by krzyk** [comment]
@simonjenga as for `{@inheritDoc}` you can get rid of those, and change the checkstyle excludes to:
```
// @checkstyle SomeExclude (100 lines)
```
Just before method declaration, it will work there.

**Comment by krzyk** [comment]
@simonjenga as for the fields, you can access them using methods, but they can't expose internal implementation.
Let's assume that you have a field to represent a `port`, you would name this field `prt`, and provide a method `port()` that will return this port as an int - this is OK because int is passed as a copy.
In case of objects you would first need to create a copy (for lists it would be something like `new ArrayList(oldList)`) or make sure the object is immutable (like `String`).

I don't know which fields you need to access, but try with the above.

**Comment by simonjenga** [comment]
@krzyk I had originally implemented this solution with your suggested approach in the first attempt and it was rejected by the Architect. That is why in this second attempt, i put both subclasses as static inside BaseFacade in order to eliminate the need for ```getters```

https://github.com/yegor256/s3auth/pull/297#issuecomment-97934969
https://github.com/yegor256/s3auth/pull/297#issuecomment-100288152

The problem we have is that we do not know what the Architect's ideal solution is for the design is!

It's better if you could request for a merge in order to know if we are right from his review. It may also help to prevent lot's of rework.

**Comment by krzyk** [comment]
@simonjenga let's do it this way, I will ask for reassignment of this ticket (as I don't like the design and proposing solutions would waste both our times) so someone else will have a look at it

**Comment by krzyk** [comment]
@davvd please assign someone else

**Comment by davvd** [comment]
> @davvd please assign someone else

@krzyk -15 points to your rating

**Comment by davvd** [comment]
> @davvd please assign someone else

@krzyk no problem, I will try to find somebody else

**Comment by simonjenga** [comment]
@krzyk I fully understand you. Some of these tickets have stringent coding requirements which you may not like or are not used to. Hope will meet in some other PR.

**Comment by davvd** [comment]
@pinaf review this one, please

**Comment by pinaf** [comment]
@davvd please assign someone else

**Comment by pinaf** [comment]
@simonjenga sorry, I don't think I can help here

**Comment by yegor256** [comment]
@simonjenga you see, they are afraid of your big pull request :) I would recommend you to drop this one and create a smaller one, with less changes and more focused ones. It will be easier to review. Also, keep this article in mind: http://www.yegor256.com/2015/06/22/valid-reasons-to-reject-bug-fix.html and

**Comment by davvd** [comment]
> @davvd please assign someone else

@pinaf 15 points was deducted from your rating

**Comment by davvd** [comment]
> @davvd please assign someone else

@pinaf right, I will find someone else, no problem

**Comment by simonjenga** [comment]
@yegor256 First it was about the design (No ```getters/setters```); and that was addressed even with a lack of communication. Now it's about a big! PR.

And mind you, they are not afraid of a about a big! PR. It's about the design requirements with this issue that ultimately leads to more time constraints while making tradeoffs. Issues that are a quick buck get reviewed rather fast :) 

Issue #286 is not small either and that is why this PR has so many changes to eliminate the duplication that is already there. I agree, changes in the ```javadocs``` should not have been done but now it's a little too late to undo them. Is that a reason enough to close this PR? Let me know...

I don't see how this PR can be made any smaller than it already is while fixing #286

**Comment by davvd** [comment]
@caarlos0 please review, thanks

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 25]
```diff
@@ -0,0 +1,427 @@
...
+     * Protected constructor.
+     * @param frontthreads The number of front threads
+     * @param frontprefix The name of the front threads
+     * @param backthreads The number of back threads
+     * @param backprefix The name of the back threads
+     * @param svr The server socket
+     * @param secure The secure server socket
+     * @checkstyle ParameterNumber (5 lines)
+     */
+    protected BaseFacade(final int frontthreads, final String frontprefix,
```
Comment (caarlos0) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga you could create an object with the number of threads and their name...

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java [hunk 26]
```diff
@@ -0,0 +1,427 @@
...
+    }
+
+    /**
+     * Process one server socket.
+     * @param svr The server socket
+     */
+    private void process(final ServerSocket svr) {
+        final Socket socket;
+        try {
+            socket = svr.accept();
```
Comment (caarlos0) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga I think you forgot to close the socker

**Comment by caarlos0** [comment]
@simonjenga some notes, @krzyk did a great job reviewing so there is not much left for me...

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java (hunk 26 above)
Comment (simonjenga) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: addressed the issue

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java (hunk 25 above)
Comment (simonjenga) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @caarlos0 Once we create this object, it would mean we also have some ```getters``` in order to retrieve the number of threads and their name so as to pass them to this method ```Executors.newScheduledThreadPool(int corePoolSize, ThreadFactory threadFactory)```. We are restricted not to use ```getters```.

**Comment by simonjenga** [comment]
@caarlos0 I think it's better to leave the ```{@inheritDoc}``` tags as they are since suppressing the associated qulice warnings as instructed above by @krzyk leaves the method to look as follows which is more verbose as it currently is.
```
// @checkstyle RegexpMultilineCheck (100 lines)
// @checkstyle RegexpSinglelineCheck (100 lines)
/**
  * @param socket The socket to report to
  * @throws IOException If something wrong happens inside
  */ 
```

File: s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java (hunk 25 above)
Comment (caarlos0) [inline s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java]: @simonjenga weird, but OK..

**Comment by caarlos0** [comment]
@simonjenga no, you can remove the entire javadoc, so the javadoc from the parent method will be used. You can do this for all `@Override` methods..

**Comment by simonjenga** [comment]
@caarlos0 You're right, removing the entire ```javadoc``` is the best thing to do.
See the addressed code review changes.

**Comment by caarlos0** [comment]
@simonjenga nice, thanks!

**Comment by caarlos0** [comment]
@rultor merge

**Comment by caarlos0** [comment]
ping @yegor256

**Comment by caarlos0** [comment]
ping @yegor256

**Comment by caarlos0** [comment]
@davvd this task seems abandoned...

**Comment by yegor256** [comment]
@caarlos0 ok, closing

**Comment by davvd** [comment]
> @davvd this task seems abandoned...

@caarlos0 @yegor256 will close it

**Comment by caarlos0** [comment]
@davvd closed already

**Comment by davvd** [comment]
> @davvd closed already

@caarlos0 thanks, got it

**Comment by davvd** [comment]
@caarlos0 thanks, paid, **23 mins** to your account, payment ID is `AP-29E52249P8937502L`, task took 1152 hours and 5 mins to complete

you're getting extra minutes here (c=8)

+23 to your rating, your total score is [+1388](http://www.netbout.com/b/35095?open=rating)
